### PR TITLE
feat: enable or disable tracing with system params

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Follow-up of #14220. Help #14468. Related #11514.

Add a system parameter `enable_tracing` (default `false`) to allow toggling the opentelemetry tracing during runtime. It's achieved by [reloading](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/reload/index.html) the `filter` layer on the opentelemetry layer.

- Since this is node-independent, a `CommonHandler` is introduced and reused by meta and all other nodes subscribing to the changes on the system params.
- The system param is only used to toggle the functionality, but not to specify the endpoint URL. This is because in cloud deployment, we may want to provision users with the endpoint and not anticipate them changing it.

Env var `RW_DISABLE_EMBEDDED_TRACING` is also removed, as the same effect can be done by setting the system param.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Document this PR together with #14220:

Introduce distributed tracing. Disabled by default. Can be turned on or off for the whole cluster with `system.enable_tracing` config or altering `enable_tracing` system parameter during runtime.

By default, the traces are exported to the meta node. Users can visualize it in the dashboard at `meta-node:5691`. It's also possible to export the traces to an external opentelemetry gRPC endpoint, by setting the env var `RW_TRACING_ENDPOINT` when launching the cluster.

### CODE REVIEW

File: src/common/src/system_param/common.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Mutex;
+
+use super::reader::SystemParamsReader;
+use crate::util::tracing::layer::toggle_otel_layer;
+
+/// Node-independent handler for system parameter changes.
```
Comment (xxchan) [inline src/common/src/system_param/common.rs:20]: What does "Node-independent handler" mean? What's node-dependent?

File: src/common/src/system_param/reader.rs [hunk 2]
```diff
@@ -77,7 +77,15 @@ impl SystemParamsReader {
...
-        self.prost.pause_on_next_bootstrap.unwrap_or(false)
+        self.prost
+            .pause_on_next_bootstrap
+            .unwrap_or_else(|| default::pause_on_next_bootstrap().unwrap())
+    }
+
+    pub fn enable_tracing(&self) -> bool {
+        self.prost
+            .enable_tracing
+            .unwrap_or_else(|| default::enable_tracing().unwrap())
```
Comment (xxchan) [inline src/common/src/system_param/reader.rs:88]:
```suggestion
            .unwrap_or_else(|| default::pause_on_next_bootstrap().unwrap_or_default())
    }

    pub fn enable_tracing(&self) -> bool {
        self.prost
            .enable_tracing
            .unwrap_or_else(|| default::enable_tracing().unwrap_or_default())
```
?

File: src/utils/runtime/src/logger.rs [hunk 3]
```diff
@@ -388,7 +394,7 @@ pub fn init_risingwave_logger(settings: LoggerSettings) {
     // Tracing layer
     #[cfg(not(madsim))]
     if let Some(endpoint) = settings.tracing_endpoint {
-        println!("tracing enabled, exported to `{endpoint}`");
+        println!("opentelemetry tracing will be exported to `{endpoint}` if enabled");
```
Comment (fuyufjh) [inline src/utils/runtime/src/logger.rs:397]: Why not `tracing::log`? I think `tracing::log` will work no matter `enable_tracing` is true or false

**Review by fuyufjh — APPROVED** [review]
LGTM

File: src/common/src/system_param/local_manager.rs [hunk 4]
```diff
@@ -40,9 +41,23 @@ pub struct LocalSystemParamsManager {
...
+
+        // Spawn a task to run the common handler.
+        tokio::spawn({
+            let mut rx = tx.subscribe();
+            async move {
+                let handler = CommonHandler::new(initial_params);
+
+                while rx.changed().await.is_ok() {
+                    let new_params = (**rx.borrow_and_update().load()).clone();
+                    handler.handle_change(new_params);
```
Comment (xxchan) [inline src/common/src/system_param/local_manager.rs:56]: Why do we need this in 3 places

**Review by xxchan — COMMENTED** [review]
Generally LGTM

File: src/utils/runtime/src/logger.rs (hunk 3 above)
Comment (BugenZhao) [reply src/utils/runtime/src/logger.rs:397]: `tracing-subscriber` is not inited until the very last line of this function. 🤣 No logs will be printed with `tracing::event` before that.

File: src/common/src/system_param/common.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/system_param/common.rs:20]: Some system param subscribers are only responsible for some specific business of that node.

- Compute:
https://github.com/risingwavelabs/risingwave/blob/99fa11570ccc58dd02899927d291b59aa2363cf5/src/compute/src/server.rs#L294-L298

- Meta:
https://github.com/risingwavelabs/risingwave/blob/99fa11570ccc58dd02899927d291b59aa2363cf5/src/meta/src/barrier/mod.rs#L737-L740

File: src/common/src/system_param/local_manager.rs (hunk 4 above)
Comment (BugenZhao) [reply src/common/src/system_param/local_manager.rs:56]: The code paths for subscribing to system parameter changes on meta nodes and other nodes are completely different. 😄 Further refactoring is needed.

File: src/common/src/system_param/reader.rs (hunk 2 above)
Comment (BugenZhao) [reply src/common/src/system_param/reader.rs:88]: They're always `Some` according to system-param definitions.

**Review by yezizp2012 — APPROVED** [review]
LGTM

File: src/common/src/system_param/reader.rs (hunk 2 above)
Comment (BugenZhao) [reply src/common/src/system_param/reader.rs:88]: Since they're generated, we may also add a new function returning the unwrapped type.

**Pull request merged**